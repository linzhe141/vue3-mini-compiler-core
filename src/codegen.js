export function generate(ast, options) {
  const context = createCodegenContext(ast, options)
  const { mode, push } = context
  // const helpers = Array.from(ast.helpers)
  // const hasHelpers = helpers.length > 0

  if (mode === 'module') {
    genModulePreamble(ast, context)
  }
  const functionName = `render`
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')
  push(`function ${functionName}(${signature}) {`)
  push(`return `)
  if (ast.codegenNode) {
    genNode(ast.codegenNode, context)
  } else {
    push(`null`)
  }
  push(`}`)
  return {
    ast,
    code: context.code,
  }
}

function isString(val) {
  return typeof val === 'string'
}
function genNullableArgs(args) {
  return args.map(arg => arg || `null`)
}
export const isArray = Array.isArray

function genNodeListAsArray(nodes, context) {
  context.push(`[`)
  genNodeList(nodes, context)
  context.push(`]`)
}
function genNodeList(nodes, context) {
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isString(node)) {
      push(node)
    } else if (isArray(node)) {
      genNodeListAsArray(node, context)
    } else {
      genNode(node, context)
    }
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}
function genVNodeCall(node, context) {
  const { push, helper } = context
  const { tag, props, children, isBlock } = node
  if (isBlock) {
    push(`(${helper('OPEN_BLOCK')}(), `)
  }
  const callHelper = isBlock ? 'CREATE_ELEMENT_BLOCK' : 'CREATE_ELEMENT_VNODE'
  push(helper(callHelper) + `(`)
  genNodeList(genNullableArgs([tag, props, children]), context)
  push(`)`)
  if (isBlock) {
    push(`)`)
  }
}
function genText(node, context) {
  context.push(JSON.stringify(node.content))
}
function genExpression(node, context) {
  // isStatic 为true 就是 <div id="13" /> 的id就是一个字符串
  // isStatic 为false 就是 <div @click="()=>foo()" /> 的click应该是一个表达式，所有在generate时，不用在进行字符串化
  const { content, isStatic } = node
  context.push(isStatic ? JSON.stringify(content) : content)
}

function genInterpolation(node, context) {
  const { push, helper } = context
  push(`${helper('TO_DISPLAY_STRING')}(`)
  genNode(node.content, context)
  push(`)`)
}
function genCompoundExpression(node, context) {
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i]
    if (isString(child)) {
      context.push(child)
    } else {
      genNode(child, context)
    }
  }
}
function genCallExpression(node, context) {
  const { push, helper } = context
  const callee = helper(node.callee)
  push(callee + `(`)
  genNodeList(node.arguments, context)
  push(`)`)
}
function genExpressionAsPropertyKey(node, context) {
  const { push } = context
  const text = node.content
  push(text)
}
function genObjectExpression(node, context) {
  const { push } = context
  const { properties } = node
  if (!properties.length) {
    push(`{}`)
    return
  }
  push(`{ `)
  for (let i = 0; i < properties.length; i++) {
    const { key, value } = properties[i]
    // key
    genExpressionAsPropertyKey(key, context)
    push(`: `)
    // value
    genNode(value, context)
    if (i < properties.length - 1) {
      push(`,`)
    }
  }
  push(` }`)
}

function genNode(node, context) {
  // <template>xx</template>
  // 这种就只push xx
  if (isString(node)) {
    context.push(node)
    return
  }
  switch (node.type) {
    case 'ELEMENT':
      genNode(node.codegenNode, context)
      break
    case 'TEXT':
      genText(node, context)
      break
    case 'SIMPLE_EXPRESSION':
      genExpression(node, context)
      break

    case 'INTERPOLATION':
      genInterpolation(node, context)
      break
    case 'TEXT_CALL':
      genNode(node.codegenNode, context)
      break
    case 'COMPOUND_EXPRESSION':
      genCompoundExpression(node, context)
      break
    case 'VNODE_CALL':
      genVNodeCall(node, context)
      break

    case 'JS_CALL_EXPRESSION':
      genCallExpression(node, context)
      break
    case 'JS_OBJECT_EXPRESSION':
      genObjectExpression(node, context)
      break
    default:
      break
  }
}

function genModulePreamble(ast, context) {
  const { push } = context
  // generate import statements for helpers
  if (ast.helpers.size) {
    const helpers = Array.from(ast.helpers)
    push(
      `import {${helpers
        .map(s => `${formatKey(s)} as _${formatKey(s)}`)
        .join(', ')}} from ${JSON.stringify('vue')}\n`
    )
  }
  push(`export `)
}

function createCodegenContext(ast, options) {
  const { mode } = options
  const context = {
    mode,
    code: ``,
    helper(key) {
      return `_${formatKey(key)}`
    },
    push(code) {
      context.code += code
    },
  }
  return context
}
function formatKey(inputString) {
  return inputString.toLowerCase().replace(/_[a-z]/g, function (match) {
    return match.charAt(1).toUpperCase()
  })
}
