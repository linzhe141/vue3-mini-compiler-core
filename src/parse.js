import { Tokenizer, isWhitespace } from './token.js'
let currentOpenTag = null
let currentProp = null
let currentAttrValue = ''

let currentInput = ''
let root = null
const stack = []

function addNode(node) {
  ;(stack[0] || root).children.push(node)
}
function endOpenTag(end) {
  addNode(currentOpenTag)
  stack.unshift(currentOpenTag)
  currentOpenTag = null
}

const tokenizer = new Tokenizer(stack, {
  ontext(start, end) {
    const content = currentInput.slice(start, end)
    console.log('ontext->', content)
    const parent = stack[0] || root
    const lastNode = parent.children[parent.children.length - 1]
    if (lastNode?.type === 'TEXT') {
      // merge 字符串拼接
      lastNode.content += content
    } else {
      parent.children.push({
        type: 'TEXT',
        content,
      })
    }
  },
  ondirname(start, end) {
    const raw = currentInput.slice(start, end)
    // @
    const name = 'on'
    // const name =
    //   raw === '.' || raw === ':'
    //     ? 'bind'
    //     : raw === '@'
    //       ? 'on'
    //       : raw === '#'
    //         ? 'slot'
    //         : raw.slice(2)

    currentProp = {
      type: 'DIRECTIVE',
      name,
      rawName: raw,
      exp: undefined,
      arg: undefined,
    }
  },
  oninterpolation(start, end) {
    let innerStart = start + '{{'.length
    let innerEnd = end - '}}'.length
    // 去除头尾空格
    while (isWhitespace(currentInput.charAt(innerStart))) {
      innerStart++
    }
    while (isWhitespace(currentInput.charAt(innerEnd - 1))) {
      innerEnd--
    }
    const exp = currentInput.slice(innerStart, innerEnd)
    addNode({
      type: 'INTERPOLATION',
      // content: createExp(exp, false, getLoc(innerStart, innerEnd)),
      content: {
        type: 'SIMPLE_EXPRESSION',
        content: exp,
        constType: 'NOT_CONSTANT',
      },
    })
  },
  ondirarg(start, end) {
    if (start === end) return
    const arg = currentInput.slice(start, end)
    currentProp.arg = {
      type: 'SIMPLE_EXPRESSION',
      content: arg,
      constType: 'CAN_STRINGIFY',
    }
  },
  onopentagname(start, end) {
    const name = currentInput.slice(start, end)
    console.log('onopentagname->', name)
    currentOpenTag = {
      type: 'ELEMENT',
      tag: name,
      props: [],
      children: [],
      // codegenNode: undefined,
    }
  },
  onopentagend(end) {
    console.log('onopentagend')
    endOpenTag(end)
  },
  onselfclosingtag(end) {
    const name = currentOpenTag.tag
    currentOpenTag.isSelfClosing = true
    endOpenTag(end)
    if (stack[0]?.tag === name) {
      stack.shift()
    }
  },
  onattribname(start, end) {
    const name = currentInput.slice(start, end)
    console.log('onattribname->', name)
    currentProp = {
      type: 'ATTRIBUTE',
      name,
      value: undefined,
    }
  },
  onattribnameend(end) {
    console.log('onattribnameend')
  },
  onattribend(quote, end) {
    console.log('onattribend')
    if (currentOpenTag && currentProp) {
      if (quote !== 'QuoteType.NoValue') {
        if (currentProp.type === 'ATTRIBUTE') {
          // assign value

          currentProp.value = {
            type: 'TEXT',
            content: currentAttrValue,
          }
        } else {
          // directive
          currentProp.exp = {
            type: 'SIMPLE_EXPRESSION',
            content: currentAttrValue,
            constType: 'NOT_CONSTANT',
          }
        }
      }
      currentOpenTag.props.push(currentProp)
    }
    currentAttrValue = ''
  },
  onattribdata(start, end) {
    const name = currentInput.slice(start, end)
    console.log('onattribdata->', name)
    currentAttrValue += name
  },
  onclosetag(start, end) {
    const name = currentInput.slice(start, end)
    console.log('onclosetag->', name)
    for (let i = 0; i < stack.length; i++) {
      const e = stack[i]
      if (e.tag.toLowerCase() === name.toLowerCase()) {
        for (let j = 0; j <= i; j++) {
          stack.shift()
          // const el = stack.shift()
          // onCloseTag(el, end, j < i)
        }
        break
      }
    }
  },
})

function createRoot(children, source) {
  return {
    type: 'ROOT',
    source,
    children,
    // codegenNode,
  }
}
export function parse(input) {
  currentInput = input
  root = createRoot([], currentInput)
  tokenizer.parse(currentInput)
  return root
}
