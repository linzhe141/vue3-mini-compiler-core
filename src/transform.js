export function transform(root, options) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root, context)
  root.helpers = new Set([...context.helpers.keys()])
}

function createRootCodegen(root, context) {
  const { helper } = context
  const { children } = root
  if (children.length === 1) {
    const child = children[0]
    root.codegenNode = child
  } else {
    // FRAGMENT
  }
}

function createTransformContext(root, options) {
  const context = {
    root,
    helpers: new Map(),
    nodeTransforms: options.nodeTransforms || [],
    helper(name) {
      context.helpers.set(name, 1)
    },
  }
  return context
}

function traverseNode(root, context) {
  const node = root
  const { nodeTransforms } = context
  const exitFns = []
  for (let i = 0; i < nodeTransforms.length; i++) {
    const onExit = nodeTransforms[i](node, context)
    if (onExit) {
      exitFns.push(onExit)
    }
  }

  switch (node.type) {
    case 'INTERPOLATION':
      context.helper('TO_DISPLAY_STRING')
      break
    case 'ELEMENT':
    case 'ROOT':
      traverseChildren(node, context)
      break
  }
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }
}

function traverseChildren(parent, context) {
  for (let i = 0; i < parent.children.length; i++) {
    const child = parent.children[i]
    traverseNode(child, context)
  }
}
