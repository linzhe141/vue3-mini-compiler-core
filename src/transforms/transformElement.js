import {
  createObjectProperty,
  createSimpleExpression,
  createObjectExpression,
} from '../ast.js'
export function transformElement(node, context) {
  return function postTransformElement() {
    if (!(node.type === 'ELEMENT')) {
      return
    }
    const { tag, props } = node
    const vnodeTag = `"${tag}"`
    let vnodeProps
    let vnodeChildren

    let shouldUseBlock = false
    // props
    if (props.length > 0) {
      const propsBuildResult = buildProps(node)
      vnodeProps = propsBuildResult.props
    }
    // children
    if (node.children.length > 0) {
      if (node.children.length === 1) {
        const child = node.children[0]
        const type = child.type
        const hasDynamicTextChild = type === 'INTERPOLATION'
        if (hasDynamicTextChild || type === 'TEXT') {
          vnodeChildren = child
        } else {
          vnodeChildren = node.children
        }
      } else {
        vnodeChildren = node.children
      }
    }
    if (shouldUseBlock) {
      context.helper('OPEN_BLOCK')
      context.helper('CREATE_ELEMENT_BLOCK')
    } else {
      context.helper('CREATE_ELEMENT_VNODE')
    }
    // if (directives) {
    //   context.helper(WITH_DIRECTIVES)
    // }
    node.codegenNode = {
      type: 'VNODE_CALL',
      tag: vnodeTag,
      props: vnodeProps,
      children: vnodeChildren,
      isBlock: shouldUseBlock,
    }
  }
}

function buildProps({ props }) {
  const properties = []
  for (let i = 0; i < props.length; i++) {
    // static attribute
    const prop = props[i]
    if (prop.type === 'ATTRIBUTE') {
      const { name, value } = prop
      let isStatic = true

      properties.push(
        createObjectProperty(
          createSimpleExpression(name, true),
          createSimpleExpression(value ? value.content : '', isStatic)
        )
      )
    } else {
      // TODO
      // directives v-on
    }
  }
  const propsExpression = createObjectExpression(properties)
  return {
    props: propsExpression,
  }
}
