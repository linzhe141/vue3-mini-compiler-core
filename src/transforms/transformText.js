export function transformText(node, context) {
  if (node.type === 'ROOT' || node.type === 'ELEMENT') {
    return () => {
      const children = node.children
      let currentContainer
      let hasText = false
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child)) {
          hasText = true
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j]
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: 'COMPOUND_EXPRESSION',
                  children: [child],
                }
              }
              currentContainer.children.push(` + `, next)
              children.splice(j, 1)
              j--
            } else {
              currentContainer = undefined
              break
            }
          }
        }
      }

      // 预先转换为 createTextVNode(text)
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child) || child.type === 'COMPOUND_EXPRESSION') {
          const callArgs = []
          if (child.type !== 'TEXT' || child.content !== ' ') {
            callArgs.push(child)
          }
          children[i] = {
            type: 'TEXT_CALL',
            content: child,
            codegenNode: {
              type: 'JS_CALL_EXPRESSION',
              callee: context.helper('CREATE_TEXT'),
              arguments: callArgs,
            },
          }
        }
      }
    }
  }
}

function isText(node) {
  return node.type === 'TEXT' || node.type === 'INTERPOLATION'
}
