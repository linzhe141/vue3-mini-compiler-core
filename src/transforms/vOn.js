import { createObjectProperty, createSimpleExpression } from '../ast.js'

export function transformOn(dir, node, context) {
  const { arg } = dir
  let eventName
  if (arg.type === 'SIMPLE_EXPRESSION') {
    let rawName = arg.content
    const eventString = 'on' + capitalize(rawName)
    eventName = createSimpleExpression(eventString, true)
  }

  const exp = dir.exp
  const ret = { props: [createObjectProperty(eventName, exp)] }
  return ret
}
const capitalize = str => str.charAt(0).toUpperCase() + str.slice(1)
