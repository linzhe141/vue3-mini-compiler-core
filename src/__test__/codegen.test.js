import { expect, describe, test } from 'vitest'
import { baseParse } from '../parse'
import { transform } from '../transform'
import { transformElement } from '../transforms/transformElement'
import { transformText } from '../transforms/transformText'
import { generate } from '../codegen'
describe('codegen', () => {
  test('element', () => {
    const input = '<div>linzhe</div>'
    const ast = baseParse(input)
    transform(ast, {
      nodeTransforms: [transformElement, transformText],
    })
    const node = generate(ast, { mode: 'module' })
    expect(node).toMatchSnapshot()
  })
})
