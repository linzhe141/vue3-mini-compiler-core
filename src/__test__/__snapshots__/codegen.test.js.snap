// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > element 1`] = `
{
  "ast": {
    "children": [
      {
        "children": [
          {
            "codegenNode": {
              "arguments": [
                {
                  "content": "linzhe",
                  "type": "TEXT",
                },
              ],
              "callee": "CREATE_TEXT",
              "type": "JS_CALL_EXPRESSION",
            },
            "content": {
              "content": "linzhe",
              "type": "TEXT",
            },
            "type": "TEXT_CALL",
          },
        ],
        "codegenNode": {
          "children": [
            {
              "codegenNode": {
                "arguments": [
                  {
                    "content": "linzhe",
                    "type": "TEXT",
                  },
                ],
                "callee": "CREATE_TEXT",
                "type": "JS_CALL_EXPRESSION",
              },
              "content": {
                "content": "linzhe",
                "type": "TEXT",
              },
              "type": "TEXT_CALL",
            },
          ],
          "isBlock": true,
          "props": undefined,
          "tag": ""div"",
          "type": "VNODE_CALL",
        },
        "props": [],
        "tag": "div",
        "type": "ELEMENT",
      },
    ],
    "codegenNode": {
      "children": [
        {
          "codegenNode": {
            "arguments": [
              {
                "content": "linzhe",
                "type": "TEXT",
              },
            ],
            "callee": "CREATE_TEXT",
            "type": "JS_CALL_EXPRESSION",
          },
          "content": {
            "content": "linzhe",
            "type": "TEXT",
          },
          "type": "TEXT_CALL",
        },
      ],
      "isBlock": true,
      "props": undefined,
      "tag": ""div"",
      "type": "VNODE_CALL",
    },
    "helpers": Set {
      "CREATE_TEXT",
      "CREATE_ELEMENT_VNODE",
      "OPEN_BLOCK",
      "CREATE_ELEMENT_BLOCK",
    },
    "source": "<div>linzhe</div>",
    "type": "ROOT",
  },
  "code": "import {createText as _createText, createElementVnode as _createElementVnode, openBlock as _openBlock, createElementBlock as _createElementBlock} from "vue"
export function render(_ctx, _cache) {return (_openBlock(), _createElementBlock("div", null, [_createText("linzhe")]))}",
}
`;
